server/src/main/java/org/elasticsearch/action/support/TimeoutReleasableListener.java

```diff
@@ -0,0 +1,112 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * A listener for a {@link Releasable} which races against a timeout, but which does not leak any {@link Releasable} with which it is
+ * eventually completed.
+ */
+public class TimeoutReleasableListener implements ActionListener<Releasable> {
+
+    private final Supplier<ActionListener<Releasable>> listenerSupplier;
+    private final Runnable timeoutCanceller;
+
+    public static TimeoutReleasableListener create(
+        ActionListener<Releasable> delegate,
+        TimeValue timeout,
+        ThreadPool threadPool,
+        String timeoutExecutor
+    ) {
+        final var listenerSupplier = readOnce(delegate);
+        return new TimeoutReleasableListener(listenerSupplier, scheduleTimeout(timeout, threadPool, timeoutExecutor, listenerSupplier));
+    }
+
+    private static Supplier<ActionListener<Releasable>> readOnce(ActionListener<Releasable> listener) {
+        final var listenerRef = new AtomicReference<>(ActionListener.assertOnce(listener));
+        return new Supplier<>() {
+            @Override
+            public ActionListener<Releasable> get() {
+                return listenerRef.getAndSet(null);
+            }
+
+            @Override
+            public String toString() {
+                return Objects.toString(listenerRef.get());
+            }
+        };
+    }
+
+    private static Runnable scheduleTimeout(
+        TimeValue timeout,
+        ThreadPool threadPool,
+        String timeoutExecutor,
+        Supplier<ActionListener<Releasable>> listenerSupplier
+    ) {
+        try {
+            final var cancellable = threadPool.schedule(() -> {
+                final var listener = listenerSupplier.get();
+                if (listener != null) {
+                    listener.onFailure(new ElasticsearchTimeoutException("timed out after [" + timeout + "/" + timeout.millis() + "ms]"));
+                }
+            }, timeout, timeoutExecutor);
+            return () -> {
+                try {
+                    cancellable.cancel();
+                } catch (Exception e) {
+                    // should not happen, we cannot reasonably do anything with an exception here anyway
+                    assert false : e;
+                }
+            };
+        } catch (Exception e) {
+            listenerSupplier.get().onFailure(e);
+            return () -> {};
```

Comment:
I wonder about this. I suppose it only happens during shutdown? Can we perhaps assert that?

In case we do get here, it seems odd that we try to block operations afterwards. Probably not a big deal if it only happens on shutdown.

To some extent we could just throw here? Or change `TimeoutReleasableListener.create` into a `run` method accepting the code to run under timeout (a `Consumer<ActionListener`), so we can skip calling it when scheduling the timeout fails?